### ISSUE TITLE
Add a URL-shortener-like service to myterms.fyi

### ISSUE THREAD

**Issue by solvaholic** [solvaholic/myterms.fyi#18]
One could use any URL shortening service to conveniently link their terms. For example, just now I created http://bit.ly/3GMOBzE as a link to [the current version of my profile README on GitHub.com](https://github.com/solvaholic/solvaholic/tree/7ce0d63eca2face8ce862fae78b03066c968f58b).

However, I want what I'm linking to be more apparent to the reader. In a digital signature, for example, I might like to include `https://myterms.fyi/<slug>` as a link to the specific terms under which I applied that digital signature.

Make it so!

**Comment by solvaholic** [comment]
Absent other compelling suggestions I'll stand this service up as an Azure Function.

Why would I start with Azure? Convenience/laziness. I'm already maintaining other domains with Azure; The interfaces Azure provides are intuitive to me; I [find](https://duckduckgo.com/?q=azure+function+URL+shortener) lots of helpful prior art to learn from.

If you see an option you think would be better, please do share!

**Comment by solvaholic** [comment]
What requirements? idk how this part of developing an application is expected to go. Happy to learn tho. Just now I think these are important:

- The URL redirection service can be deployed easily using code in this repository
- The service is reliable, and inexpensive to operate
- The service consistently returns a redirect in, idk, 2 seconds or less
- The relationship between a myterms.fyi URL and the URL it links to is transparent and onymous
- It's simple for a user to obtain information about a myterms.fyi URL without following it

Other thoughts?

**Comment by solvaholic** [comment]
This is vague, and money is squishy, so I want to put numbers around it:

> inexpensive to operate

I plan to set this up under my personal Azure account. I'm willing to fund setup and the first 3 months operating costs, up to USD $100 total. I'll post billing statements in this repository and we can discuss how to pay ongoing costs.

Why $100? It should cost nowhere near that, more like $5 or $10 per month. I set my limit higher because I see this as a valuable learning opportunity and don't want cost to block progress.

**Comment by solvaholic** [comment]
> I [find](https://duckduckgo.com/?q=azure+function+URL+shortener) lots of helpful prior art to learn from.

I tried using https://github.com/FBoucher/AzUrlShortener today and didn't get it working. I tried without the front-end. After installing the app in my Azure account I followed https://github.com/FBoucher/AzUrlShortener/blob/main/doc/howto-use-azure-storage-explorer.md to Azure Storage Explorer and did not see a UrlsDetails table.

I'll read through these, follow questions to documentation, and see how to proceed:
- https://wp.sjkp.dk/building-a-url-shortener-with-azure-functions-in-less-than-100-lines-of-code/
- https://medium.com/marcus-tee-anytime/create-your-own-url-shortener-host-in-azure-almost-free-for-cloud-infrastructure-a74c9cc29720
- https://github.com/estiller/azure-functions-url-shortener
- https://github.com/hanzlamateen/azure-function-url-shortener

**Comment by solvaholic** [comment]
tl;dr? Use an API manager.

🤔 I had in mind an Azure Function would differentiate request methods and paths in the URL. For example I thought one function would handle each of these requests, including authentication for the PUT and DELETE methods:
- `PUT myterms.fyi/<slug>?url=<url>` creates or replaces a slug:URL mapping
- `GET myterms.fyi/<slug>` returns [HTTP 302](https://developer.mozilla.org/docs/Web/HTTP/Status/302), redirecting to the slug's URL
- `GET myterms.fyi/<slug>.info` returns, or redirects to, a page about the slug
- `DELETE myterms.fyi/<slug>` deletes a slug:URL mapping

Having read through the links in https://github.com/solvaholic/myterms.fyi/issues/18#issuecomment-1328102831 I think I should use an API Manager in front of one or more functions:
https://learn.microsoft.com/en-us/azure/api-management/

That way authentication can remain separate from the function(s), and each function can be simpler, faster, and easier to read and maintain. This approach could also speed progress toward a useful product:
- `GET myterms.fyi/<slug>` returns [HTTP 302](https://developer.mozilla.org/docs/Web/HTTP/Status/302), redirecting to the slug's URL

**Comment by solvaholic** [comment]
Looking back to the examples in https://github.com/solvaholic/myterms.fyi/issues/18#issuecomment-1328102831 I think I don't want to learn how to use another programming language. But 😆 wait, I'm not a programmer - do I really know how to use any programming language? [C#](https://learn.microsoft.com/en-us/dotnet/csharp/tour-of-csharp/tutorials/) intrigues me and seems like a popular option; [JavaScript](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Introduction) is at least a little familiar; [Go](https://golangbot.com/golang-tutorial-part-1-introduction-and-installation/) has been on my to-learn list a while.

I'm already experienced in reading and learning languages, and have less practice learning the scaffolding and conventions developers use. So, if I'm going to maintain this code*, I'll look through several Azure Functions examples and pick a language that seems intuitive to me.

\* I don't especially want to maintain code for a URL redirector. If I can learn how to re-use one, I may prefer to go that way.

**Comment by letsnotsay** [comment]
It is sad but appears true that only the very young and the retired can afford to devote the hours needed to master any of these systems deeply - in my own practice, its the 20 percent of anything I use 80% of the time that gets attention, the balance remaining buried in the swamp of deep learning.

No wonder all the "smart phones" are hopelessly junked up and root is often hard to obtain, the investment in time almost guarantees few will ever wander deep enough down those rabbit holes to get in a position to eject the bloat and seize control of the hardware :)

Sounds like you have a good understanding of "how to" implement this crazy little idea.  Now if we can find a well-meaning gazillionaire to donate a few hundred million to the appropriate powers that be, we could get some paid for lawmaker to insert a rule someplace (thinking CFR) requiring any company affecting commerce between the States by mandating mandatory arbitration for their customers to pay a fee earmarked to fund our little defensive terms of service service. We could start a non-profit NGO thing to run it all and start peeling back the abusive practice of arbitration with our own abusive practice of TOS injection LOL... think vaccination against the theft of the right to meaningful recourse... mandated by the same overbearing government that has mandated the theft to begin with! 

Seems like the [current] American Way in a twisted sort of way.

Perhaps for our next plot to better humanity we could find a way to insure against insurance companies that refuse to pay lawful claims so that there's money for the screwed insureds to secure counsel and sue the little facist panties off the culprits. Sorry, I better stop now, its too early for such rich fantasies just yet... Thanks for your thoughts, never thought about Azure, but for all my disdain of things MS, it does seem to run things much better than expected.

**Comment by solvaholic** [comment]
Aye, @letsnotsay, thank you for stopping :wink: I think it's never too early to fantasize. And I think this issue can be resolved without addressing those problems.

> I'll look through several Azure Functions examples and pick a language that seems intuitive to me.

Eventually I followed
https://github.com/Azure-Samples/todo-csharp-sql-swa-func

to
https://learn.microsoft.com/en-us/azure/developher/azure-developer-cli/azd-templates?tabs=java#choose-a-template

With `azd` I can compare the same application template in 3 different languages:

```text
% azd template list | grep Functions
Azure-Samples/todo-csharp-sql-swa-func     ToDo Application with a C# API and Azure SQL Databse on Static Web Apps and Functions
Azure-Samples/todo-nodejs-mongo-swa-func   ToDo Application with a Node.js API and Azure Cosmos DB API for MongoDB on Static Web Apps and Functions
Azure-Samples/todo-python-mongo-swa-func   ToDo Application with a Python API and Azure Cosmos DB API for MongoDB on Static Web Apps and Functions
```

Which I thought would be a straightforward way to see the differences. However, `azd up`, or the templates, seem to assume that I have a local development environment set up with prerequisites for each project. I don't. And I don't want to install a bunch of software I plan to not use.

"I'll run this in containers," I thought. Didn't get that going, either, but I did experience https://github.com/devcontainers a little.

If I'm going to invest more time and attention, I'll need to know it's doing more for me than helping me see what an application looks like in three different languages. I'll go back to write-ups like the ones linked in https://github.com/solvaholic/myterms.fyi/issues/18#issuecomment-1328102831, follow along with what I have on hand, and see how that goes.

**Comment by solvaholic** [comment]
> https://medium.com/marcus-tee-anytime/create-your-own-url-shortener-host-in-azure-almost-free-for-cloud-infrastructure-a74c9cc29720

I think Marcus Tee followed these steps:

1. Code the URL shortener
1. Set up the database
1. Create the function app
1. Create the API manager
1. Configure authorization with managed identities
1. Create the front-end application
1. Package everything up in a Git repository
1. Automate deployment

Looking back to requirements listed in https://github.com/solvaholic/myterms.fyi/issues/18#issuecomment-1327690379, I think Marcus' example meets these already:

> * The URL redirection service can be deployed easily using code in this repository
> * The service is reliable, and inexpensive to operate
> * The service consistently returns a redirect in, idk, 2 seconds or less

And I think these could be added easily enough:

> * The relationship between a myterms.fyi URL and the URL it links to is transparent and onymous
> * It's simple for a user to obtain information about a myterms.fyi URL without following it

If people will use this application, we'll want to know the data are backed up and restore-able. We'll also want to know what data are stored, how they're used, and how any personal or sensitive data are protected. Those are important but don't need to block this issue, so I opened new issues for them:
[What are myterms.fyi's terms? #19](https://github.com/solvaholic/myterms.fyi/issues/19)
[Add database backup and restore to myterms.fyi #20](https://github.com/solvaholic/myterms.fyi/issues/20)
